reactos/drivers/usb/usbhub_new/pnp.c

```diff
@@ -0,0 +1,2822 @@
+
+            PdoDevice->Flags |= DO_POWER_PAGABLE;
+            PdoDevice->Flags &= ~DO_DEVICE_INITIALIZING;
+
+            DeviceRelations->Objects[DeviceRelations->Count++] = PdoDevice;
+
+            PdoExtension = PdoDevice->DeviceExtension;
+            PdoExtension->PortPdoFlags &= ~USBHUB_PDO_FLAG_POWER_D1_OR_D2;
+
+            continue;
+        }
+
+        USBH_Wait(100);
+
+        NtStatus = USBH_SyncResetPort(HubExtension, Port);
+
+        if (!NT_SUCCESS(NtStatus))
+        {
+            if (HubExtension->HubFlags & USBHUB_FDO_FLAG_USB20_HUB)
+            {
+                PortData->DeviceObject = NULL;
+                PortData->ConnectionStatus = NoDeviceConnected;
+                continue;
+            }
+        }
+        else
+        {
+            NtStatus = USBH_SyncGetPortStatus(HubExtension,
+                                              Port,
+                                              &PortData->PortStatus,
+                                              sizeof(USBHUB_PORT_STATUS));
+
+            UsbPortStatus = PortData->PortStatus.UsbPortStatus;
+
+            if (NT_SUCCESS(NtStatus))
+            {
+                ULONG ix = 0;
+
+                for (NtStatus = USBH_CreateDevice(HubExtension, Port, UsbPortStatus, ix);
+                     !NT_SUCCESS(NtStatus);
+                     NtStatus = USBH_CreateDevice(HubExtension, Port, UsbPortStatus, ix))
+                {
+                    USBH_Wait(500);
+
+                    if (ix >= 2)
+                    {
+                        break;
+                    }
+
+                    if (PortData->DeviceObject)
+                    {
+                        IoDeleteDevice(PortData->DeviceObject);
+                        PortData->DeviceObject = NULL;
+                        PortData->ConnectionStatus = NoDeviceConnected;
+                    }
+
+                    USBH_SyncResetPort(HubExtension, Port);
+
+                    ix++;
+                }
+
+                if (NT_SUCCESS(NtStatus))
+                {
+                    if (PortData->DeviceObject)
+                    {
+                        PdoExtension = PortData->DeviceObject->DeviceExtension;
+
+                        if (!(PdoExtension->PortPdoFlags & USBHUB_PDO_FLAG_PORT_LOW_SPEED) &&
+                            !(PdoExtension->PortPdoFlags & USBHUB_PDO_FLAG_PORT_HIGH_SPEED) &&
+                            !(HubExtension->HubFlags & USBHUB_FDO_FLAG_USB20_HUB))
+                        {
+                            DPRINT1("USBH_FdoQueryBusRelations: FIXME USBH_DeviceIs2xDualMode()\n");
+
+                            if (0)//USBH_DeviceIs2xDualMode(PdoExtension))
+                            {
+                                PdoExtension->PortPdoFlags |= USBHUB_PDO_FLAG_HS_USB1_DUALMODE;
+                            }
+                        }
+                    }
+
+                    ObReferenceObject(PortData->DeviceObject);
```

Comment:
pnp.c:1286: You check whether `PortData->DeviceObject` is NULL above (line 1269), yet here (and below in line 1290/1291) you dereference it outside a null check.